baybe/recommenders/pure/bayesian/botorch.py

```diff
@@ -151,31 +170,174 @@ def _recommend_continuous(
+                subspace_continuous,
+                batch_size,
+            )
+
+        # Return optimized points as dataframe
+        rec = pd.DataFrame(points, columns=subspace_continuous.param_names)
+        return rec
+
+    def _recommend_continuous_with_cardinality_constraints(
+        self,
+        subspace_continuous: SubspaceContinuous,
+        batch_size: int,
+    ) -> tuple[Tensor, Tensor]:
+        """Recommend from a continuous search space with cardinality constraints.
+
+        Args:
+            subspace_continuous: The continuous subspace from which to generate
+                recommendations.
+            batch_size: The size of the recommendation batch.
+
+        Returns:
+            The recommendations.
+            The acquisition values.
+
+        Raises:
+            RuntimeError: If the continuous search space has no cardinality constraint.
+        """
+        import torch
+
+        if not subspace_continuous.constraints_cardinality:
+            raise RuntimeError(
+                f"This method expects a subspace object with constraints of type "
+                f"{ContinuousCardinalityConstraint.__name__}. For a subspace object "
+                f"without constraints of type"
+                f" {ContinuousCardinalityConstraint.__name__}, use method"
+                f"{self._recommend_continuous_without_cardinality_constraints.__name__}."  # noqa
+            )
+
+        acqf_values_all: list[Tensor] = []
+        points_all: list[Tensor] = []
+
+        def append_recommendation_for_inactive_parameters_setting(
+            inactive_parameters: Collection[str],
+        ):
+            """Append the recommendation for each inactive parameter configuration.
+
+            Args:
+                inactive_parameters: A list of inactive parameters.
+            """
+            # Create a new subspace by ensuring all active parameters being
+            # non-zeros.
+            subspace_continuous_without_cardinality_constraints = (
+                subspace_continuous._remove_cardinality_constraints(inactive_parameters)
+            )
+            try:
+                # Optimize the acquisition function
+                (
+                    points_i,
+                    acqf_values_i,
+                ) = self._recommend_continuous_without_cardinality_constraints(
+                    subspace_continuous_without_cardinality_constraints,
+                    batch_size,
+                )
+                # Append recommendation list and acquisition function values
+                points_all.append(points_i.unsqueeze(0))
+                acqf_values_all.append(acqf_values_i.unsqueeze(0))
+
+            # The optimization problem may be infeasible for certain inactive
+            # parameters. The optimize_acqf raises a ValueError when the optimization
+            # problem is infeasible.
+            except ValueError:
```

Comment:
since `ValueError` is extremely common also for all kinds of other errors, should this be amde more precise via message?

Reply:
can you elaborate what you suggest in detail?

Reply:
simply a custom infeasibility error

Reply:
I think you got the context wrong: we're not `raising` the ValueError here – we're catching the one that botorch raises and simply skip the corresponding subspace because it's infeasible. It's the part we discussed yesterday where ideally botorch would raise an infeasibility error itself to avoid ambiguity with other errors (i.e numerical ones). But we can't raise anything here.

Reply:
I will create a Botorch issue/feature request on customised Error for infeasibility later and let you know.

Reply:
@Waschenbacher saw your feature request. Just wanted to encourage you to draft a PR for this yourself if this is really a minor change. I provided two PRs myself just recently, and the guys are really helpful and it was quite straight-forward to just create a PR and get it merged. I think it took roughly a day until everything was reviewed and merged from their end.
Also, you need to sign a CLA (or similar), and I've already got the confirmation from our legal colleagues that we are officially allowed to sign this ;)

Reply:
@AVHopp Good point. The guys approached me and asked for a PR :) . They would need it at several other positions as well. Will create a PR and let you know when it gets merged. I will take a look what is a CLA. Thanks for the insights!

Reply:
The CLA is basically saying "Although I wrote some code of this package, I give you all of the rights to it" - and this needs to be somehow "signed" by us. Just wanted to already tell you "You can do this, don't worry", but if you have more questions, just ask :)